### PR TITLE
Add prerender indicator for app router

### PR DESCRIPTION
**Pull request by ijjk**
## Background

Currently in app router during development all pages are dynamic rendered and streamed since we don't know if a page will be prerendered until after the page is rendered. During build time we do a render pass for each path and if it doesn't bail from static generation it is considered prerendered/ISR'd. 

Since it can be tricky to track all the bail outs for static generation and doing numerous builds to track this isn't feasible we want to have some sort of signal in development to show if it would have been prerendered. 

## New Handling

To alleviate the above pain point we are planning to bring back the prerendered indicator from pages initially and have a config to disable if desired. We are also going to investigate other possible indicators potentially through headers or similar. 

<details>

<summary>Preview</summary>

CleanShot 2024-06-29 at 08 56 11@2x

CleanShot 2024-06-29 at 08 55 46@2x

</details>

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 30f75e7a28268737721e1542acde4ed3eb8b3b8b

`TURBOPACK=1 pnpm test test/integration/auto-export/test/index.test.js` (turbopack)
- Auto Export > dev > should not show hydration warning from mismatching asPath

<details>
<summary>Expand output</summary>

  ● Auto Export › dev › should not show hydration warning from mismatching asPath

    expect(received).toEqual(expected) // deep equality

    - Expected  -  1
    + Received  + 10

    - Array []
    + Array [
    +   Array [
    +     "[HMR] Invalid message: [object Object]
    + Error: Unexpected action app-isr-manifest
    +     at http://localhost:38089/_next/static/chunks/%5Broot%20of%20the%20server%5D__fea3e8._.js:19774:27
    +     at processMessage (http://localhost:38089/_next/static/chunks/%5Broot%20of%20the%20server%5D__fea3e8._.js:19591:21)
    +     at http://localhost:38089/_next/static/chunks/%5Broot%20of%20the%20server%5D__fea3e8._.js:19349:13
    +     at WebSocket.handleMessage (http://localhost:38089/_next/static/chunks/%5Broot%20of%20the%20server%5D__fea3e8._.js:9241:17)",
    +   ],
    + ]

      88 |       const browser = await webdriver(appPort, '/zeit/cmnt-1')
      89 |       const caughtWarns = await browser.eval(`window.caughtWarns`)
    > 90 |       expect(caughtWarns).toEqual([])
         |                           ^
      91 |     })
      92 |
      93 |     it('should include error link when hydration error does occur', async () => {

      at Object.toEqual (integration/auto-export/test/index.test.js:90:27)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| buildDuration | 23.8s | 20.1s | N/A |
| buildDurationCached | 11.2s | 9.9s | N/A |
| nodeModulesSize | 359 MB | 359 MB | ⚠️ +146 kB |
| nextStartRea..uration (ms) | 493ms | 519ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| 1780.HASH.js gzip | 167 B | 167 B | ✓ |
| 5453-HASH.js gzip | 36.1 kB | 37.6 kB | ⚠️ +1.52 kB |
| 7514-HASH.js gzip | 5.07 kB | 5.06 kB | N/A |
| a7a62840-HASH.js gzip | 51.7 kB | 51.7 kB | N/A |
| framework-HASH.js gzip | 56.7 kB | 56.7 kB | N/A |
| main-app-HASH.js gzip | 221 B | 222 B | N/A |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.71 kB | 1.71 kB | ✓ |
| Overall change | 37.9 kB | 39.5 kB | ⚠️ +1.52 kB |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 193 B | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 509 B | 510 B | N/A |
| css-HASH.js gzip | 341 B | 341 B | ✓ |
| dynamic-HASH.js gzip | 2.52 kB | 2.52 kB | ✓ |
| edge-ssr-HASH.js gzip | 264 B | 266 B | N/A |
| head-HASH.js gzip | 362 B | 363 B | N/A |
| hooks-HASH.js gzip | 391 B | 390 B | N/A |
| image-HASH.js gzip | 4.27 kB | 4.27 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.69 kB | 2.69 kB | N/A |
| routerDirect..HASH.js gzip | 326 B | 325 B | N/A |
| script-HASH.js gzip | 396 B | 397 B | N/A |
| withRouter-HASH.js gzip | 320 B | 321 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.62 kB | 3.62 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| _buildManifest.js gzip | 485 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| index.html gzip | 522 B | 524 B | N/A |
| link.html gzip | 537 B | 539 B | N/A |
| withRouter.html gzip | 519 B | 520 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| edge-ssr.js gzip | 127 kB | 127 kB | N/A |
| page.js gzip | 167 kB | 167 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 660 B | 657 B | N/A |
| middleware-r..fest.js gzip | 156 B | 155 B | N/A |
| middleware.js gzip | 29.5 kB | 29.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.03 kB | 1.03 kB | ✓ |
| Overall change | 1.03 kB | 1.03 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 184 kB | 184 kB | ⚠️ +102 B |
| app-page-exp..prod.js gzip | 112 kB | 112 kB | N/A |
| app-page-tur..prod.js gzip | 123 kB | 123 kB | N/A |
| app-page-tur..prod.js gzip | 119 kB | 119 kB | N/A |
| app-page.run...dev.js gzip | 178 kB | 178 kB | ⚠️ +102 B |
| app-page.run..prod.js gzip | 109 kB | 109 kB | N/A |
| app-route-ex...dev.js gzip | 23.4 kB | 23.4 kB | ✓ |
| app-route-ex..prod.js gzip | 18.8 kB | 18.8 kB | ✓ |
| app-route-tu..prod.js gzip | 18.8 kB | 18.8 kB | ✓ |
| app-route-tu..prod.js gzip | 18.6 kB | 18.6 kB | ✓ |
| app-route.ru...dev.js gzip | 24.6 kB | 24.6 kB | ✓ |
| app-route.ru..prod.js gzip | 18.6 kB | 18.6 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 21.6 kB | 21.6 kB | ✓ |
| pages.runtim...dev.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim..prod.js gzip | 21.6 kB | 21.6 kB | ✓ |
| server.runti..prod.js gzip | 51.7 kB | 51.7 kB | N/A |
| Overall change | 579 kB | 579 kB | ⚠️ +204 B |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js ijjk/app-isr-indicator | Change |
| - | - | - | - |
| 0.pack gzip | 1.67 MB | 1.67 MB | ⚠️ +2.31 kB |
| index.pack gzip | 128 kB | 129 kB | ⚠️ +473 B |
| Overall change | 1.8 MB | 1.8 MB | ⚠️ +2.78 kB |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [8358],
   {
-    /***/ 1362: /***/ (
+    /***/ 9618: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -9,7 +9,7 @@
       (window.__NEXT_P = window.__NEXT_P || []).push([
         "/image",
         function () {
-          return __webpack_require__(2160);
+          return __webpack_require__(699);
         },
       ]);
       if (false) {
@@ -18,7 +18,7 @@
       /***/
     },
 
-    /***/ 537: /***/ (module, exports, __webpack_require__) => {
+    /***/ 9451: /***/ (module, exports, __webpack_require__) => {
       "use strict";
       /* __next_internal_client_entry_do_not_use__  cjs */
       Object.defineProperty(exports, "__esModule", {
@@ -40,15 +40,15 @@
         __webpack_require__(3537)
       );
       const _head = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(7092)
+        __webpack_require__(6490)
       );
-      const _getimgprops = __webpack_require__(9834);
-      const _imageconfig = __webpack_require__(5676);
-      const _imageconfigcontextsharedruntime = __webpack_require__(387);
-      const _warnonce = __webpack_require__(451);
-      const _routercontextsharedruntime = __webpack_require__(5357);
+      const _getimgprops = __webpack_require__(3646);
+      const _imageconfig = __webpack_require__(535);
+      const _imageconfigcontextsharedruntime = __webpack_require__(4724);
+      const _warnonce = __webpack_require__(6321);
+      const _routercontextsharedruntime = __webpack_require__(1759);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(3945)
+        __webpack_require__(1882)
       );
       // This is replaced by webpack define plugin
       const configEnv = {
@@ -376,7 +376,7 @@
       /***/
     },
 
-    /***/ 9834: /***/ (
+    /***/ 3646: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -392,9 +392,9 @@
           return getImgProps;
         },
       });
-      const _warnonce = __webpack_require__(451);
-      const _imageblursvg = __webpack_require__(3547);
-      const _imageconfig = __webpack_require__(5676);
+      const _warnonce = __webpack_require__(6321);
+      const _imageblursvg = __webpack_require__(8297);
+      const _imageconfig = __webpack_require__(535);
       const VALID_LOADING_VALUES =
         /* unused pure expression or super */ null && [
           "lazy",
@@ -769,7 +769,7 @@
       /***/
     },
 
-    /***/ 3547: /***/ (__unused_webpack_module, exports) => {
+    /***/ 8297: /***/ (__unused_webpack_module, exports) => {
       "use strict";
       /**
        * A shared function, used on both client and server, to generate a SVG blur placeholder.
@@ -824,7 +824,7 @@
       /***/
     },
 
-    /***/ 6850: /***/ (
+    /***/ 973: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -851,10 +851,10 @@
         },
       });
       const _interop_require_default = __webpack_require__(1478);
-      const _getimgprops = __webpack_require__(9834);
-      const _imagecomponent = __webpack_require__(537);
+      const _getimgprops = __webpack_require__(3646);
+      const _imagecomponent = __webpack_require__(9451);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(3945)
+        __webpack_require__(1882)
       );
       function getImageProps(imgProps) {
         const { props } = (0, _getimgprops.getImgProps)(imgProps, {
@@ -886,7 +886,7 @@
       /***/
     },
 
-    /***/ 3945: /***/ (__unused_webpack_module, exports) => {
+    /***/ 1882: /***/ (__unused_webpack_module, exports) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -921,7 +921,7 @@
       /***/
     },
 
-    /***/ 2160: /***/ (
+    /***/ 699: /***/ (
       __unused_webpack_module,
       __webpack_exports__,
       __webpack_require__
@@ -938,8 +938,8 @@
 
       // EXTERNAL MODULE: ./node_modules/.pnpm/react@19.0.0-rc-6230622a1a-20240610/node_modules/react/jsx-runtime.js
       var jsx_runtime = __webpack_require__(898);
-      // EXTERNAL MODULE: ./node_modules/.pnpm/next@file+..+main-repo+packages+next+next-packed.tgz_react-dom@19.0.0-rc-6230622a1a-20240610__zehaskxadtwcczqqbmt6koh6bq/node_modules/next/image.js
-      var next_image = __webpack_require__(6793);
+      // EXTERNAL MODULE: ./node_modules/.pnpm/next@file+..+diff-repo+packages+next+next-packed.tgz_react-dom@19.0.0-rc-6230622a1a-20240610__uulzbengwsfwhwaa2ambxampcy/node_modules/next/image.js
+      var next_image = __webpack_require__(1428);
       var image_default = /*#__PURE__*/ __webpack_require__.n(next_image); // CONCATENATED MODULE: ./pages/nextjs.png
       /* harmony default export */ const nextjs = {
         src: "/_next/static/media/nextjs.cae0b805.png",
@@ -969,12 +969,12 @@
       /***/
     },
 
-    /***/ 6793: /***/ (
+    /***/ 1428: /***/ (
       module,
       __unused_webpack_exports,
       __webpack_require__
     ) => {
-      module.exports = __webpack_require__(6850);
+      module.exports = __webpack_require__(973);
 
       /***/
     },
@@ -984,7 +984,7 @@
     /******/ var __webpack_exec__ = (moduleId) =>
       __webpack_require__((__webpack_require__.s = moduleId));
     /******/ __webpack_require__.O(0, [2888, 9774, 179], () =>
-      __webpack_exec__(1362)
+      __webpack_exec__(9618)
     );
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
<details>
<summary>Diff for <strong>5453-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 30f75e7a28268737721e1542acde4ed3eb8b3b8b

**Comment by ijjk** [comment]
## Failing test suites 
Commit: e2a57209c344f5159b053e13a99312bb73be6c3f

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/app-dir/actions-navigation/index.test.ts` (PPR)
- app-dir action handling > should handle actions correctly after following a relative link

<details>
<summary>Expand output</summary>

  ● app-dir action handling › should handle actions correctly after following a relative link

    elementHandle.click: Timeout 60000ms exceeded.
    Call log:
      - attempting click action
      -   waiting for element to be visible, enabled and stable
      -     element is not visible - waiting...

      394 |   click() {
      395 |     return this.chain((el: ElementHandleExt) => {
    > 396 |       return el.click().then(() => el)
          |                 ^
      397 |     })
      398 |   }
      399 |

      at click (lib/browsers/playwright.ts:396:17)
      at Object.<anonymous> (e2e/app-dir/actions-navigation/index.test.ts:41:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: e2a57209c344f5159b053e13a99312bb73be6c3f

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/app-dir/actions-navigation/index.test.ts` (PPR)
- app-dir action handling > should handle actions correctly after following a relative link

<details>
<summary>Expand output</summary>

  ● app-dir action handling › should handle actions correctly after following a relative link

    elementHandle.click: Timeout 60000ms exceeded.
    Call log:
      - attempting click action
      -   waiting for element to be visible, enabled and stable
      -     element is not visible - waiting...

      394 |   click() {
      395 |     return this.chain((el: ElementHandleExt) => {
    > 396 |       return el.click().then(() => el)
          |                 ^
      397 |     })
      398 |   }
      399 |

      at click (lib/browsers/playwright.ts:396:17)
      at Object.<anonymous> (e2e/app-dir/actions-navigation/index.test.ts:41:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 805646b6ec03d018b0051b9c73d320686309a102

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/app-dir/actions-navigation/index.test.ts` (PPR)
- app-dir action handling > should handle actions correctly after following a relative link

<details>
<summary>Expand output</summary>

  ● app-dir action handling › should handle actions correctly after following a relative link

    elementHandle.click: Timeout 60000ms exceeded.
    Call log:
      - attempting click action
      -   waiting for element to be visible, enabled and stable
      -     element is not visible - waiting...

      394 |   click() {
      395 |     return this.chain((el: ElementHandleExt) => {
    > 396 |       return el.click().then(() => el)
          |                 ^
      397 |     })
      398 |   }
      399 |

      at click (lib/browsers/playwright.ts:396:17)
      at Object.<anonymous> (e2e/app-dir/actions-navigation/index.test.ts:41:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 378d9c4a673b8c6056dc798a9bb05fcd770a610a

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/app-dir/actions-navigation/index.test.ts` (PPR)
- app-dir action handling > should handle actions correctly after following a relative link

<details>
<summary>Expand output</summary>

  ● app-dir action handling › should handle actions correctly after following a relative link

    elementHandle.click: Timeout 60000ms exceeded.
    Call log:
      - attempting click action
      -   waiting for element to be visible, enabled and stable
      -     element is not visible - waiting...

      394 |   click() {
      395 |     return this.chain((el: ElementHandleExt) => {
    > 396 |       return el.click().then(() => el)
          |                 ^
      397 |     })
      398 |   }
      399 |

      at click (lib/browsers/playwright.ts:396:17)
      at Object.<anonymous> (e2e/app-dir/actions-navigation/index.test.ts:41:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 378d9c4a673b8c6056dc798a9bb05fcd770a610a

`TURBOPACK=1 pnpm test-start test/e2e/app-dir/actions-navigation/index.test.ts` (turbopack)
- app-dir action handling > should handle actions correctly after following a relative link

<details>
<summary>Expand output</summary>

  ● app-dir action handling › should handle actions correctly after following a relative link

    thrown: "Exceeded timeout of 60000 ms for a test.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      34 |   })
      35 |
    > 36 |   it('should handle actions correctly after following a relative link', async () => {
         |   ^
      37 |     const browser = await next.browser('/nested-folder/products')
      38 |
      39 |     await browser.elementByCss('a').click()

      at it (e2e/app-dir/actions-navigation/index.test.ts:36:3)
      at Object.describe (e2e/app-dir/actions-navigation/index.test.ts:4:1)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 62a3dc1c159131dd479a6719e0cb52e77d0bec40

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/app-dir/actions-navigation/index.test.ts` (PPR)
- app-dir action handling > should handle actions correctly after following a relative link

<details>
<summary>Expand output</summary>

  ● app-dir action handling › should handle actions correctly after following a relative link

    elementHandle.click: Timeout 60000ms exceeded.
    Call log:
      - attempting click action
      -   waiting for element to be visible, enabled and stable
      -     element is not visible - waiting...

      394 |   click() {
      395 |     return this.chain((el: ElementHandleExt) => {
    > 396 |       return el.click().then(() => el)
          |                 ^
      397 |     })
      398 |   }
      399 |

      at click (lib/browsers/playwright.ts:396:17)
      at Object.<anonymous> (e2e/app-dir/actions-navigation/index.test.ts:41:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 62a3dc1c159131dd479a6719e0cb52e77d0bec40

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/app-dir/actions-navigation/index.test.ts` (PPR)
- app-dir action handling > should handle actions correctly after following a relative link

<details>
<summary>Expand output</summary>

  ● app-dir action handling › should handle actions correctly after following a relative link

    elementHandle.click: Timeout 60000ms exceeded.
    Call log:
      - attempting click action
      -   waiting for element to be visible, enabled and stable
      -     element is not visible - waiting...

      394 |   click() {
      395 |     return this.chain((el: ElementHandleExt) => {
    > 396 |       return el.click().then(() => el)
          |                 ^
      397 |     })
      398 |   }
      399 |

      at click (lib/browsers/playwright.ts:396:17)
      at Object.<anonymous> (e2e/app-dir/actions-navigation/index.test.ts:41:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 142e224ca7564a1d6da7574b16e1450854911145

`pnpm test-dev test/e2e/app-dir/actions/app-action.test.ts`
- app-dir action handling > fetch actions > should handle redirects to routes that provide an invalid RSC response

<details>
<summary>Expand output</summary>

  ● app-dir action handling › fetch actions › should handle redirects to routes that provide an invalid RSC response

    expect(received).toContain(expected) // indexOf

    Expected substring: "Hello from a pages route"
    Received string:    "0.5680865301725937
    Client
    Server
    Client and Server
    0
    +1+1 (Slow)-1*2
    redirect to a pages route
    submit
    test"

      1035 |
      1036 |       await retry(async () => {
    > 1037 |         expect(await browser.elementByCss('body').text()).toContain(
           |                                                           ^
      1038 |           'Hello from a pages route'
      1039 |         )
      1040 |         expect(await browser.url()).toBe(`${next.url}/pages-dir`)

      at toContain (e2e/app-dir/actions/app-action.test.ts:1037:59)
      at retry (lib/next-test-utils.ts:787:14)
      at Object.<anonymous> (e2e/app-dir/actions/app-action.test.ts:1036:7)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/src/server/dev/hot-reloader-turbopack.ts [hunk 1]
```diff
@@ -544,8 +544,9 @@ export async function createHotReloaderTurbopack(
     },
 
     // TODO: Figure out if socket type can match the NextJsHotReloaderInterface
-    onHMR(req, socket: Socket, head) {
+    onHMR(req, socket: Socket, head, callback) {
```
Comment (huozhi) [inline packages/next/src/server/dev/hot-reloader-turbopack.ts:547]: callback arg (callback) in a callback (onHMR) might be confusing

```suggestion
    onHMR(req, socket: Socket, head, onUpgrade) {
```

File: packages/next/src/server/dev/hot-reloader-turbopack.ts [hunk 2]
```diff
@@ -544,8 +544,9 @@ export async function createHotReloaderTurbopack(
...
       wsServer.handleUpgrade(req, socket, head, (client) => {
+        callback(client)
```
Comment (huozhi) [inline packages/next/src/server/dev/hot-reloader-turbopack.ts:549]:
```suggestion
        onUpgrade(client)
```

File: packages/next/src/server/dev/hot-reloader-types.ts [hunk 3]
```diff
@@ -142,7 +149,12 @@ export interface NextJsHotReloaderInterface {
   stop(): Promise<void>
   send(action: HMR_ACTION_TYPES): void
   getCompilationErrors(page: string): Promise<any[]>
-  onHMR(req: IncomingMessage, _socket: Duplex, head: Buffer): void
+  onHMR(
+    req: IncomingMessage,
+    _socket: Duplex,
+    head: Buffer,
+    callback: (client: { send(data: string): void }) => void
```
Comment (huozhi) [inline packages/next/src/server/dev/hot-reloader-types.ts:156]:
```suggestion
    onUpgrade: (client: { send(data: string): void }) => void
```

**Review by huozhi — COMMENTED** [review]
lgtm in general

**Review by huozhi — APPROVED** [review]

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 53cdac1d5358abccf69bff029adaa8985f6737ae

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/app-dir/navigation/navigation.test.ts` (PPR)
- app dir - navigation > navigation between pages and app > should not contain _rsc query while navigating from app to pages

<details>
<summary>Expand output</summary>

  ● app dir - navigation › navigation between pages and app › should not contain _rsc query while navigating from app to pages

    page.waitForSelector: Timeout 60000ms exceeded.
    Call log:
      - waiting for locator('#link-to-app')

      423 |     return this.chain(() => {
      424 |       return page
    > 425 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      426 |         .then(async (el) => {
      427 |           // it seems selenium waits longer and tests rely on this behavior
      428 |           // so we wait for the load event fire before returning

      at waitForSelector (lib/browsers/playwright.ts:425:10)
      at Object.<anonymous> (e2e/app-dir/navigation/navigation.test.ts:631:7)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Pull request merged**